aptos/token_bridge/sources/bridge_state.move

```diff
@@ -118,6 +147,92 @@ module token_bridge::bridge_state {
+
+        // initialize new coin using CoinType
+        let name = bridge_structs::get_name(&_asset_meta);
+        let symbol = bridge_structs::get_symbol(&_asset_meta);
+        let decimals = bridge_structs::get_decimals(&_asset_meta);
+        let monitor_supply = false;
+        let (burn_cap, freeze_cap, mint_cap) = initialize<CoinType>(&coin_signer, utf8(name), utf8(symbol), decimals, monitor_supply);
+
+        // update the following two mappings in State
+        // 1. (native chain, native address) => wrapped address
+        // 2. wrapped address => (native chain, native address)
+        let token_address = bridge_structs::get_token_address(& _asset_meta);
+        let token_chain = bridge_structs::get_token_chain(& _asset_meta);
+        let native_info = NativeInfo {token_address: token_address, token_chain: token_chain};
+        set_native_info(token_address, &native_info);
+        set_wrapped_asset(&native_info, token_address);
+
+        // store coin capabilities
+        let _token_bridge_signer = token_bridge_signer();
+        let coin_caps = CoinCapabilities<CoinType> { mint_cap: mint_cap, freeze_cap: freeze_cap, burn_cap: burn_cap};
+        move_to(&_token_bridge_signer, coin_caps);
+
+        vaa::destroy(vaa);
+    }
+
+    // transfer a native or wraped token from sender to token_bridge
+    public fun transfer_tokens<CoinType>(_sender: &signer, _amount: u128, _arbiter_fee: u128): TransferResult acquires State {
+        // TODO: assertions and checks,
+        // check if CoinType is registered with @token_bridge
+        let token_address = hash_type_info<CoinType>();
+
+        // transfer tokens from sender to token_bridge
+        transfer<CoinType>(_sender, @token_bridge, (_amount as u64));
+
+        // return TransferResult encapsulating details of token transferred
+        let native_chain = u16::from_u64(0);
+        let native_address = vector::empty<u8>();
+
+        if (is_wrapped_asset(token_address)) {
+            let _native_info = native_info(token_address);
+            native_chain = _native_info.token_chain;
+            native_address = _native_info.token_address;
+        } else if (is_registered_native_asset(token_address)) {
+            native_chain = get_chain_id();
+            native_address = token_address;
+        } else {
+            // TODO - if unregistered asset, then register the asset in the native_asset map?
+        };
+
+        // TODO - normalize amount by using helpers from utils.move
+        let normalized_amount = u256::from_u128(_amount);
+        // TODO - normalize arbiter fee
+        let normalized_arbiter_fee = u256::from_u128(_arbiter_fee);
+        let wormhole_fee = u256::from_u64(0);
+
+        let transfer_result: TransferResult = create_transfer_result(
+            native_chain,
+            native_address,
+            normalized_amount,
+            normalized_arbiter_fee,
+            wormhole_fee
+        );
+        transfer_result
```

Comment:
this function should publish a message at the end

Reply:
this function is meant to be transfer_tokens_internal, going to make the "wrapper" function now